src/shared/utilities.js

```diff
@@ -22,3 +22,13 @@ export const turnSpacesIntoPlusses = (string) => {
 
   return string;
 }
+
+export const turnDashesIntoSpacesAndCapitalize = (string) => {
```

Comment:
So, first up, using a regex with a `global` modifier (`/.../g`) will search the whole string:

```
const newString = string.replace(/-/g, " ");
```

Next up, check this out:
```
const turnDashesIntoSpaces = (string) => string.replace(/-/g, " ")
const normalizeString = (string) => string.charAt(0).toUpperCase() + string.slice(1);
const turnDashesIntoSpacesAndCapitalize = (string) => {
  const spacedString = turnDashesIntoSpaces(string);
  const normalizedString = normalizeString(spacedString);
  return normalizedString;
}
```

Now, that's pretty verbose, but follow me on the next implementation:

```
import _flow from 'lodash/flow';

const turnDashesIntoSpaces = (string) => string.replace(/-/g, " ");
const capitalizeString = (string) => string.charAt(0).toUpperCase() + string.slice(1);
const turnDashesIntoSpacesAndCapitalize = _flow(
  turnDashesIntoSpaces,
  capitalizeString,
);
```

What `_flow` (a.k.a. `pipe` in other languages or utility libraries) does is:
- accepts an infinite number of functions as arguments
- returns a function which _composes_ all of those functions into one sequence
- in that sequence, passes the return value from one function in the sequence to the next as its input

At the end of `_flow` you get a series of functions which say exactly what they do, and you can execute that on one argument or more, depending on argument shape (a.k.a. function signature) the functions you are composing with expect.

In the case of `turnDashesIntoSpacesAndCapitalize`, each function we are composing with:
- accepts a single string
- returns a single string

This means we can predictably invoke `turnDashesIntoSpacesAndCapitalize` with a single string, and expect a single string back.

This pattern allows for some cool stuff. As a for instance:

```
const logger = (thingToLog) => {
  console.log(thingToLog);
  return thingToLog;
}
const turnDashesIntoSpaces = (string) => string.replace(/-/g, " ");
const capitalizeString = (string) => string.charAt(0).toUpperCase() + string.slice(1);
const turnDashesIntoSpacesAndCapitalize = _flow(
  turnDashesIntoSpaces,
  logger,
  capitalizeString,
  logger,
);
```

The `logger` function accepts and returns the same shape that we are using in the rest of the flow, so we can still predict the overall input and output.

Here's one more level, if you're into it:
```
const logger = (message) => (thingToLog) => {
  console.log(message, thingToLog);
  return thingToLog;
}
const turnDashesIntoSpaces = (string) => string.replace(/-/g, " ");
const capitalizeString = (string) => string.charAt(0).toUpperCase() + string.slice(1);
const turnDashesIntoSpacesAndCapitalize = _flow(
  logger('before anything'),
  turnDashesIntoSpaces,
  logger('after dashes'),
  capitalizeString,
  logger('after capitalization'),
);
```

Let's break down the definition of `logger`, which is known as a **Higher Order Function**.

```
// No short arrow functions
const logger = (message) => {
  return (thingToLog) => {
    console.log(message, thingToLog);
    return thingToLog;
  }
}
```
```
// More verbose, but perhaps more clear?
const logger = (message) => {
  const loggerWithThing = (thingToLog) => {
    console.log(message, thingToLog);
    return thingToLog;
  }
  return loggerWithThing;
}
```

The fact that we have two returns nested inside each other means that we need to invoke the function _twice_ in order to actually get it to do what we intend it to do:
```
const message = 'hello there';
const thingToLog = 'I am a thing to log to the console after the message';
logger(message)(thingToLog); // console has: 'hello there', 'I am a thing...'
```

In the original usage we demonstrated inside of the `_flow` call, we only invoked the function once, right? This gave the logger HOF the _message_, but later when the `turnDashesIntoSpacesAndCapitalize` is invoked a second time with a string that you'd like to transform.

For completeness, this pattern could be extended an infinite number of times:

```
const logger = (messageOne) => (messageTwo) => (thingToLog) => {
  console.log(messageOne, messageTwo, thingToLog);
  return thingToLog;
}
```

You can then compose a generic version ahead of time by invoking to one level, and adding more specificity as you go along:

```
const logger = (messageOne) => (messageTwo) => (thingToLog) => {
  console.log(messageOne, messageTwo, thingToLog);
  return thingToLog;
}
const beforeLogger = logger('before');
const afterLogger = logger('after');

...

const turnDashesIntoSpacesAndCapitalize = _flow(
  beforeLogger('anything'),
  turnDashesIntoSpaces,
  afterLogger('dashes'),
  capitalizeString,
  afterLogger('capitalization'),
);
```

I know this last example is a bit contrived, but I promise that it's one of the most powerful patterns I've learned in programming. It is also more generally referred to under the umbrella called **functional composition**.

Reply:
ok I mostly get it. Not 100% sure why it's better than doing it the other way, besides saving space but it seems to make sense.

Reply:
It's something that will become more clear the more you break your procedural code (like standard "inline" stuff) into small functions. I'll try and think of a better scenario to exemplify this.

Reply:
ok, thanks :)